File: GGAndroid_2.0/GGPay/pay-wxpay/src/main/java/io/ganguo/pay/wxpay/WXPayService.java
package io.ganguo.pay.wxpay;

import android.app.Activity;
import android.app.Application;

import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;

import java.lang.ref.WeakReference;

import io.ganguo.pay.core.GGPay;
import io.ganguo.pay.core.IPayService;
import io.ganguo.pay.core.PayCallback;
import io.ganguo.pay.core.PayResult;

/**
 * Created by Roger on 06/07/2017.
 */

final class WXPayService implements IPayService {
    private static final String TAG = "WXPayService";

    private WXPayData wxPayData;
    private PayCallback payCallback;
    private WeakReference<Activity> ref;
    private Application.ActivityLifecycleCallbacks lifecycleCallbacks;
    private boolean isReleased = false;

    WXPayService(Activity activity, PayCallback payCallback, WXPayData wxPayData) {
        if (payCallback == null) {
            throw new NullPointerException("pay callback can not be null");
        }
        if (activity == null) {
            throw new NullPointerException("activity can not be null");
        }

        this.wxPayData = wxPayData;
        this.payCallback = payCallback;
        this.ref = new WeakReference<Activity>(activity);
        lifecycleCallbacks = createLifecycleCallbacks();
        activity.getApplication().registerActivityLifecycleCallbacks(lifecycleCallbacks);
    }

    private Application.ActivityLifecycleCallbacks createLifecycleCallbacks() {
        return new SimpleActivityLifecycleCallbacks() {
            @Override
            public void onActivityDestroyed(Activity activity) {
                super.onActivityDestroyed(activity);
//                Log.i(TAG, "destroyed activity name: " + activity.getClass().getSimpleName());

                if (!isReleased && activity instanceof IWXAPIEventHandler) {
                    String resultCode = activity.getIntent().getStringExtra(WXConstants.WXPAY_RESULT);
                    PayResult payResult = activity.getIntent().getParcelableExtra(WXConstants.WXPAY_EXTRA);
                    if (isEquals(resultCode, WXConstants.CODE_SUCCESS)) {
                        payCallback.onPaySuccess(payResult);
                    } else if (isEquals(resultCode, WXConstants.CODE_FAIL)) {
                        payCallback.onPayError(payResult);
                    } else if (isEquals(resultCode, WXConstants.CODE_CANCEL)) {
                        payCallback.onPayCancel(payResult);
                    } else if (isEquals(resultCode, WXConstants.CODE_NO_RESPONSE)) {
                        payCallback.onPayError(payResult);
                    }
                }
                GGPay.remove(WXPayService.this);
            }
        };
    }

    /**
     * 发起支付，为避免内存泄露，pay 方法在支付回调返回后会自动释放资源，因此不能多次调用。
     */
    @Override
    public boolean pay() {
        if (isReleased) {
            throw new IllegalStateException("IPayService already release, you can not reuse a IPayService");
        }
        if (ref.get() == null) {
            return false;
        }
        final IWXAPI wxApi = WXAPIFactory.createWXAPI(ref.get(), WXConstants.getWeChatAppId());
        // 检查是否安装微信
        if (!wxApi.isWXAppInstalled()) {
            wxApi.detach();
            payCallback.onPayNotFound("wxpay");
            return false;
        }
        wxApi.registerApp(WXConstants.getWeChatAppId());

        wxApi.sendReq(wxPayData.toPayReq());
        wxApi.detach();
        return true;
    }

    @Override
    public void release() {
        synchronized (this) {
            wxPayData = null;
            payCallback = null;
            if (ref.get() != null) {
                ref.get().getApplication().unregisterActivityLifecycleCallbacks(lifecycleCallbacks);
            }
            lifecycleCallbacks = null;
            ref.clear();
            isReleased = true;
        }
    }

    @Override
    public boolean isRelease() {
        return isReleased;
    }

    private boolean isEquals(String s, String target) {
        return s != null && target != null && s.equals(target);
    }
}

File: GGAndroid_3.0/core/service/pay/pay-wxpay/src/main/java/io/ganguo/pay/wxpay/WXPayService.kt
package io.ganguo.pay.wxpay

import android.app.Activity
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.WXAPIFactory
import io.ganguo.pay.core.annotation.PayType
import io.ganguo.pay.core.exception.PayServiceException
import io.ganguo.pay.wxpay.entity.WXPayEntity
import java.lang.ref.WeakReference

/**
 * Created by Roger on 06/07/2017.
 * 微信支付服务
 * @property wxPayEntity 微信支付订单数据
 */
class WXPayService(activity: Activity?, private var wxPayEntity: WXPayEntity?) : AWXPayService() {
    private val ref: WeakReference<Activity>
    private val wxApi: IWXAPI

    init {
        if (activity == null) {
            throw NullPointerException("activity can not be null")
        }
        this.ref = WeakReference(activity)
        wxApi = WXAPIFactory.createWXAPI(ref.get(), WXConstants.WX_APP_ID)
    }

    /**
     * 发起支付，为避免内存泄露，pay 方法在支付回调返回后会自动释放资源，因此不能多次调用。
     * @return
     */
    override fun startService(): WXPayResultObservable {
        if (ref.get() == null) {
            return resultObserver
        }
        // 检查是否安装微信
        if (!wxApi.isWXAppInstalled) {
            wxApi.detach()
            return resultObserver
        }
        wxApi.registerApp(WXConstants.WX_APP_ID)
        wxApi.sendReq(wxPayEntity!!.toPayReq())
        wxApi.detach()
        return resultObserver
    }

    /**
     * 检测微信支付SDK配置是否正确
     * @return Throwable?
     */
    override fun checkException(): Throwable? {
        var payServiceException: PayServiceException? = null
        //appId为空时，抛异常
        if (!WXConstants.WX_SDK_IS_INIT) {
            val msg = "Please initialize WeChat sdk in Application!!!"
            val errorCode = WXConstants.ErrorCode.WX_PAY_SDK_NOT_INIT
            payServiceException = PayServiceException(PayType.WX_PAY, errorCode, msg)
        }
        //app未安装时，抛运行时异常
        else if (!wxApi.isWXAppInstalled) {
            val msg = "Please install the weChat first!!!"
            val errorCode = WXConstants.ErrorCode.WX_PAY_APP_NOT_INSTALLED
            payServiceException = PayServiceException(PayType.WX_PAY, errorCode, msg)
        }
        return payServiceException
    }


    /**
     * 资源释放。
     */
    override fun release() {
        super.release()
        wxPayEntity = null
        ref.clear()
    }


}
